File: frontend/src/api/batchNodeGeneration.ts
/** Client API génération batch multi-parents FR88. */
import { apiClient } from './client'

export const BATCH_GENERATE_JOB_MIN = 10

export type BatchGenerateParentStatus = 'ok' | 'error' | 'skipped' | 'cancelled'

export type BatchGenerateSuggestedConnection = {
  from?: string
  to?: string
  from_node?: string
  to_node?: string
  via_choice_index?: number | null
  connection_type?: string
}

export type BatchGenerateParentItem = {
  parent_node_id: string
  status: BatchGenerateParentStatus
  nodes: Array<Record<string, unknown> & { id: string }>
  suggested_connections: BatchGenerateSuggestedConnection[]
  error?: string | null
  warning?: string | null
  generated_choices_count?: number | null
  connected_choices_count?: number | null
  failed_choices_count?: number | null
  total_choices_count?: number | null
  context_gdd_content_fingerprint?: string | null
}

export type BatchGenerateReport = {
  items: BatchGenerateParentItem[]
  cancelled: boolean
  started_at: string
  finished_at: string
  ok_count: number
  error_count: number
  skipped_count: number
  total_nodes_generated: number
}

export type BatchGenerateJobStatus = {
  job_id: string
  status: 'queued' | 'running' | 'completed' | 'cancelled' | 'error'
  current: number
  total: number
  detail: string
  cancelled: boolean
  error?: string | null
  report?: BatchGenerateReport | null
}

export type BatchGenerateParentSpec = {
  parent_node_id: string
  parent_node_content: Record<string, unknown>
  context_selections: Record<string, unknown>
  user_instructions?: string
  npc_speaker_id?: string
  player_character_id?: string
}

export type StartBatchGenerateJobRequest = {
  document_id?: string
  parents: BatchGenerateParentSpec[]
  dialogue_nodes?: Array<Record<string, unknown>>
  llm_model_identifier?: string
  system_prompt_override?: string
  max_choices?: number | null
  choices_mode?: 'free' | 'capped'
}

export async function startBatchGenerateFromNodesJob(
  body: StartBatchGenerateJobRequest
): Promise<{ job_id: string; status: string; total: number }> {
  const response = await apiClient.post<{
    job_id: string
    status: string
    total: number
  }>('/api/v1/unity-dialogues/graph/batch-generate-from-nodes/jobs', body, {
    timeout: 60000,
  })
  return response.data
}

export async function getBatchGenerateFromNodesJob(
  jobId: string
): Promise<BatchGenerateJobStatus> {
  const response = await apiClient.get<BatchGenerateJobStatus>(
    `/api/v1/unity-dialogues/graph/batch-generate-from-nodes/jobs/${jobId}`
  )
  return response.data
}

export async function cancelBatchGenerateFromNodesJob(
  jobId: string
): Promise<BatchGenerateJobStatus> {
  const response = await apiClient.post<BatchGenerateJobStatus>(
    `/api/v1/unity-dialogues/graph/batch-generate-from-nodes/jobs/${jobId}/cancel`
  )
  return response.data
}

File: api/schemas/batch_node_generation.py
"""Schémas Pydantic — génération batch multi-parents (Story 8.9 / FR88)."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class BatchGenerateParentSpec(BaseModel):
    """Un nœud parent à traiter dans le lot."""

    parent_node_id: str = Field(..., description="ID du nœud parent")
    parent_node_content: Dict[str, Any] = Field(
        ..., description="Contenu du nœud parent (speaker, line, choices…)"
    )
    context_selections: Dict[str, Any] = Field(
        default_factory=dict,
        description="Snapshot GDD du nœud ou contexte global",
    )
    user_instructions: str = Field(
        default="",
        description="Instructions utilisateur (défaut backend si vide)",
    )
    npc_speaker_id: Optional[str] = Field(None, description="ID PNJ")
    player_character_id: Optional[str] = Field(None, description="ID PJ")


class BatchGenerateFromNodesJobRequest(BaseModel):
    """Corps POST pour démarrer un job de génération batch (N ≥ 10)."""

    document_id: Optional[str] = Field(
        None, description="ID dialogue pour annotation coûts"
    )
    parents: List[BatchGenerateParentSpec] = Field(
        ..., min_length=2, description="Parents sélectionnés (≥ 2)"
    )
    dialogue_nodes: Optional[List[Dict[str, Any]]] = Field(
        None, description="Snapshot graphe pour chemin ancêtre"
    )
    llm_model_identifier: Optional[str] = Field(None, description="Modèle LLM")
    system_prompt_override: Optional[str] = None
    max_choices: Optional[int] = None
    choices_mode: Optional[Literal["free", "capped"]] = None


class SuggestedConnectionPayload(BaseModel):
    """Connexion suggérée (alias from/to pour le front)."""

    model_config = ConfigDict(populate_by_name=True, serialize_by_alias=True)

    from_node: str = Field(..., alias="from")
    to_node: str = Field(..., alias="to")
    via_choice_index: Optional[int] = None
    connection_type: str = "choice"


class BatchGenerateParentItemResponse(BaseModel):
    """Résultat par parent."""

    parent_node_id: str
    status: Literal["ok", "error", "skipped", "cancelled"]
    nodes: List[Dict[str, Any]] = Field(default_factory=list)
    suggested_connections: List[SuggestedConnectionPayload] = Field(
        default_factory=list
    )
    error: Optional[str] = None
    warning: Optional[str] = None
    generated_choices_count: Optional[int] = None
    connected_choices_count: Optional[int] = None
    failed_choices_count: Optional[int] = None
    total_choices_count: Optional[int] = None
    context_gdd_content_fingerprint: Optional[str] = None


class BatchGenerateFromNodesReport(BaseModel):
    """Rapport agrégé d'un lot."""

    items: List[BatchGenerateParentItemResponse]
    cancelled: bool = False
    started_at: str = ""
    finished_at: str = ""
    ok_count: int = 0
    error_count: int = 0
    skipped_count: int = 0
    total_nodes_generated: int = 0


class BatchGenerateJobCreateResponse(BaseModel):
    """Réponse 202 à la création du job."""

    job_id: str
    status: Literal["queued"] = "queued"
    total: int


class BatchGenerateJobStatusResponse(BaseModel):
    """État / rapport d'un job de génération batch."""

    job_id: str
    status: Literal["queued", "running", "completed", "cancelled", "error"]
    current: int = 0
    total: int = 0
    detail: str = ""
    cancelled: bool = False
    error: Optional[str] = None
    report: Optional[BatchGenerateFromNodesReport] = None
